### PR TITLE
impelment PUT api for threshold updates

### PR DESCRIPTION
**Pull request by nhsiehgit**
PUT api for updating thresholds

### CODE REVIEW

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 1]
```diff
@@ -41,3 +62,25 @@ def get(
         self, request: Request, project: Project, release_threshold: ReleaseThreshold
     ) -> HttpResponse:
         return Response(serialize(release_threshold, request.user), status=200)
+
+    def put(
+        self, request: Request, project: Project, release_threshold: ReleaseThreshold
+    ) -> HttpResponse:
+        serializer = ReleaseThresholdPUTSerializer(data=request.data)
+
+        if serializer.is_valid():
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:71]: early return 😶
```suggestion
        if not serializer.is_valid():
            return Response(serializer.errors, status=400)
```

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 2]
```diff
@@ -41,3 +62,25 @@ def get(
...
+                kwargs["threshold_type"] = result["threshold_type"]
+            if "trigger_type" in result:
+                kwargs["trigger_type"] = result["trigger_type"]
+            if "value" in result:
+                kwargs["value"] = result["value"]
+            if "window_in_seconds" in result:
+                kwargs["window_in_seconds"] = result["window_in_seconds"]
+
+            if kwargs:
+                release_threshold.update(**kwargs)
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:84]: I'm a little confused by `kwargs` here, could you not pass result directly into the update? I think any random fields not in the serializer won't be included in `validated_data`

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 3]
```diff
@@ -11,14 +12,34 @@
...
+)
+from sentry.models.release_threshold.constants import TriggerType as ReleaseThresholdTriggerType
 from sentry.models.release_threshold.release_threshold import ReleaseThreshold
 
 
+class ReleaseThresholdPUTSerializer(serializers.Serializer):
+    threshold_type = serializers.ChoiceField(choices=ReleaseThresholdType.as_str_choices())
+    trigger_type = serializers.ChoiceField(choices=ReleaseThresholdTriggerType.as_str_choices())
+    value = serializers.IntegerField()
+    window_in_seconds = serializers.IntegerField()
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:28]: Does this have a min or max value?

File: tests/sentry/api/endpoints/release_thresholds/test_release_threshold_details.py [hunk 4]
```diff
@@ -71,3 +72,81 @@ def test_valid(self):
...
+        )
+        response = self.client.put(url, data=updated_data)
+
+        assert response.status_code == 200
+        assert response.data["id"] == str(self.basic_threshold.id)
+        assert response.data["threshold_type"] == "total_error_count"
+        assert response.data["trigger_type"] == updated_data["trigger_type"]
+        assert response.data["value"] == updated_data["value"]
+        assert response.data["window_in_seconds"] == updated_data["window_in_seconds"]
+        assert response.data["environment"]["name"] == "canary"
```
Comment (schew2381) [inline tests/sentry/api/endpoints/release_thresholds/test_release_threshold_details.py:152]:
```suggestion
        assert response.data == {
            "id": str(self.basic_threshold.id),
            "threshold_type": "total_error_count",
            "trigger_type": updated_data["trigger_type"],
            "value": updated_data["value"],
            "window_in_seconds": updated_data["window_in_seconds"],
            "environment": {"name": "canary"},
        }
```

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 5]
```diff
@@ -11,14 +12,34 @@
...
+
+    def validate_threshold_type(self, threshold_type: str):
+        return THRESHOLD_TYPE_STR_TO_INT[threshold_type]
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:31]: This is pretty cool, I didn't know you could validate and convert the input to the correct serializer field type

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 6]
```diff
@@ -11,14 +12,34 @@
...
+    def validate_trigger_type(self, threshold_type: str):
+        return TRIGGER_TYPE_STRING_TO_INT[threshold_type]
+
+
 @region_silo_endpoint
 class ReleaseThresholdDetailsEndpoint(ProjectEndpoint):
     owner: ApiOwner = ApiOwner.ENTERPRISE
     publish_status = {
         "GET": ApiPublishStatus.PRIVATE,
+        "PUT": ApiPublishStatus.PRIVATE,
```
Comment (schew2381) [inline src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:42]: Do we plan on making these public once release thresholds are out?

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py (hunk 2 above)
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:84]: hmmm...
actually you're right, that should be fine

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py (hunk 1 above)
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:71]: 🤷 
time/complexity wise this is the same.

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py (hunk 3 above)
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:28]: good call, i'll pin it to 0

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py [hunk 7]
```diff
@@ -11,14 +12,34 @@
...
+    threshold_type = serializers.ChoiceField(choices=ReleaseThresholdType.as_str_choices())
+    trigger_type = serializers.ChoiceField(choices=ReleaseThresholdTriggerType.as_str_choices())
+    value = serializers.IntegerField()
+    window_in_seconds = serializers.IntegerField()
+
+    def validate_threshold_type(self, threshold_type: str):
+        return THRESHOLD_TYPE_STR_TO_INT[threshold_type]
+
+    def validate_trigger_type(self, threshold_type: str):
+        return TRIGGER_TYPE_STRING_TO_INT[threshold_type]
```
[the remark this answers is not in the document]
Comment (sentaur-athena) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:34]: Same for threshold type.

**Review by sentaur-athena — COMMENTED** [review]
Looks good. Just a few comments.

File: src/sentry/api/endpoints/release_thresholds/release_threshold_details.py (hunk 6 above)
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/release_threshold_details.py:42]: Yup!

**Review by sentaur-athena — APPROVED** [review]

**Review by schew2381 — APPROVED** [review]
lgtm, just have to get some eyes on the migration

**Pull request merged**